narayana-spring-boot-core/src/main/java/me/snowdrop/boot/narayana/core/jdbc/AbstractXADataSourceWrapper.java

```diff
@@ -25,32 +25,37 @@
 import org.springframework.boot.jdbc.XADataSourceWrapper;
 
 /**
- * {@link XADataSourceWrapper} implementation that uses {@link NarayanaDataSource} to wrap an {@link XADataSource}.
+ * An abstract {@link XADataSourceWrapper} implementation which creates a {@link XAResourceRecoveryHelper} capable to
+ * recovery transactions of the provided {@link XADataSource} and then delegates the actual data source wrapper to its
+ * children classes.
  *
  * @author <a href="mailto:gytis@redhat.com">Gytis Trikleris</a>
  */
-public class NarayanaXADataSourceWrapper implements XADataSourceWrapper {
+public abstract class AbstractXADataSourceWrapper implements XADataSourceWrapper {
 
     private final NarayanaProperties properties;
 
     private final XARecoveryModule xaRecoveryModule;
 
-    /**
-     * Create a new {@link NarayanaXADataSourceWrapper} instance.
-     *
-     * @param properties       Narayana properties.
-     * @param xaRecoveryModule recovery module to register data source with.
-     */
-    public NarayanaXADataSourceWrapper(NarayanaProperties properties, XARecoveryModule xaRecoveryModule) {
+    protected AbstractXADataSourceWrapper(NarayanaProperties properties, XARecoveryModule xaRecoveryModule) {
         this.properties = properties;
         this.xaRecoveryModule = xaRecoveryModule;
     }
 
+    protected abstract DataSource internalWrapDataSource(XADataSource dataSource) throws Exception;
```

Comment:
I would usually place the `internal` "qualifier" at the end, but I guess it's just a matter of taste

Reply:
I'll change it

Reply:
Cool